internal/controllers/topology/cluster/reconcile_state.go

```diff
@@ -442,7 +530,7 @@ type reconcileReferencedObjectInput struct {
 // reconcileReferencedObject reconciles the desired state of the referenced object.
 // NOTE: After a referenced object is created it is assumed that the reference should
 // never change (only the content of the object can eventually change). Thus, we are checking for strict compatibility.
-func (r *Reconciler) reconcileReferencedObject(ctx context.Context, in reconcileReferencedObjectInput) error {
+func (r *Reconciler) reconcileReferencedObject(ctx context.Context, in reconcileReferencedObjectInput) (*unstructured.Unstructured, error) {
```

Comment:
I think we should avoid returning the object here and instead use current or re-read the newly created object if required. Same for reconcileReferencedTemplate

Reply:
If understood correctly we need the UID of the newly reconciled object to set an OwnerReference. Which essentially means we have to Get the object with the apiReader directly afterwards to avoid a stale cache.

I just thought we could instead just use the object returned by create which is up-to-date.

Reply:
This does not solves all the use cases, given that in some cases the owner ref is the machineDeployment.
So given that for the machine deployment we need to re-read, I prefer to keep things consistent and re-read for the control-plane as well

Reply:
The client uses a cache to read, (and it doesn't automatically update the cache on write) so there is the possibility that the MHC doesn't get created on the first reconcile using the current structure. The MHC should be created on the subsequent reconcile - which should happen very quickly on cluster creation and machineDeployment changes at least- so it will eventually get into the correct state.

Reply:
This should be fixed now